cli/pcluster/config/param_types.py

```diff
@@ -0,0 +1,1202 @@
+
+    def _check_allowed_values(self):
+        """Verify if the parameter value is one of the allowed values specified in the mapping file."""
+        allowed_values = self.map.get("allowed_values", None)
+        if allowed_values:
+            if isinstance(allowed_values, list):
+                if self.value not in allowed_values:
+                    error(
+                        "The configuration parameter '{0}' has an invalid value '{1}'\n"
+                        "Allowed values are: {2}".format(self.key, self.value, allowed_values)
+                    )
+            else:
+                # convert to regex
+                if not re.compile(allowed_values).match(str(self.value)):
+                    error(
+                        "The configuration parameter '{0}' has an invalid value '{1}'\n"
+                        "Allowed values are: {2}".format(self.key, self.value, allowed_values)
+                    )
+
+    def validate(self, fail_on_error=True):
+        """Call validation function for the parameter, if there."""
+        validation_func = self.map.get("validator", None)
+
+        if not validation_func:
+            LOGGER.debug("Configuration parameter '%s' has no validator", self.key)
+        elif self.value is None:
+            LOGGER.debug("Configuration parameter '%s' has not a value", self.key)
+        else:
+            errors, warnings = validation_func(self.key, self.value, self.pcluster_config)
+            if errors:
+                error(
+                    "The configuration parameter '{0}' has an invalid value '{1}'\n"
+                    "{2}".format(self.key, self.value, "\n".join(errors)),
+                    fail_on_error,
+                )
+            elif warnings:
+                warn(
+                    "The configuration parameter '{0}' has a wrong value '{1}'\n{2}".format(
+                        self.key, self.value, "\n".join(warnings)
+                    )
+                )
+            else:
+                LOGGER.debug("Configuration parameter '%s' is valid", self.key)
+
+    def to_file(self, config_parser):
+        """Set parameter in the config_parser in the right section."""
+        section_name = _get_file_section_name(self.section_key, self.section_label)
+        if self.value is not None and self.value != self.get_default_value():
+            config_parser.set(section_name, self.key, str(self.value))
+        else:
+            # remove parameter from config_parser if there
+            try:
+                config_parser.remove_option(section_name, self.key)
+            except NoSectionError:
+                pass
+
+    def to_cfn(self):
+        """Convert parameter to CFN representation, if "cfn" attribute is present in the Param map."""
+        cfn_params = {}
+        cfn_converter = self.map.get("cfn", None)
+
+        if cfn_converter:
+            cfn_value = self.get_cfn_value()
+            cfn_params.update({cfn_converter: str(cfn_value)})
```

Comment:
cfn_converter is simply the cfn_key right? Also we could simply do `cfn_params[cfn_converter]=str(cfn_value)` in case we always need to add a single param (same somewhere else)

Reply:
Done